### PR TITLE
Chakra and ChakraCore JSI Refactoring Part One

### PR DESCRIPTION
**Pull request by hansenyy**
This PR contains Part One of an attempt at refactoring our existing Chakra and ChakraCore JSI implementation. The goals of the refactor include:

- Use a class hierarchy instead of conditional compilation for better code sharing between Chakra and ChakraCore JSI.
- Ensure that Chakra and ChakraCore JSI do not indirectly depend on any unnecessary packages such as OpenSSL.
- Move Microsoft specific JSI augmentations (e.g. bytecode execution, promises, memory usage tracking) outside of the base implementation of Chakra and ChakraCore JSI. Instead, we use constructs in Facebook's decorator.h to implement them.

The new architecture is not yet fully implemented. Known missing areas include HostFunctions and HostObjects, as well as promises, memory usage tracking, and in-proc debugging. For now we will keep both the old and new code in the repository.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2934)

### CODE REVIEW

**Comment by acoates-ms** [comment]
FYI the `AzureDevOps - PR (UWP PR PublicRNX86Debug)` failures are around not supporting building against facebook/react-native.  React-native-windows needs to be able to build against an unforked version.  -- Using the forked versions should provide light up features, but not be required to use react-native-windows.

File: vnext/ReactWindows.sln [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29009.5
+# Visual Studio 15
+VisualStudioVersion = 15.0.28307.757
```
Comment (tudorms) [inline vnext/ReactWindows.sln]: You may want to reverse these changes.

**Comment by tudorms** [comment]
All the Chakra code shows up as brand new, making it difficult to review. It would be nice if you could make the changes in steps (start by moving / copying ChakraJsiRuntime*.h/.cpp into the new JSI/ projects, then build your new features on top).

**Comment by hansenyy** [comment]
Let me go ahead and do that. I'll send out a separate PR.

---
In reply to: [522799784](https://github.com/microsoft/react-native-windows/pull/2934#issuecomment-522799784) [](ancestors = 522799784)

**Comment by mganandraj** [comment]
> Let me go ahead and do that. I'll send out a separate PR.
> 
> In reply to: 522799784 [](ancestors = 522799784)

Hi Yichen.. Looking forward for the follow up PR !!  Thanks!

**Pull request closed without merging**